### PR TITLE
Add MultiInOutTransformer

### PR DESCRIPTION
**Pull request by austinzh**
This Transformer is added to  better support spark transformers that has inputCols and outputCols.
The rules are:
In Spark side, all parameters save/load as its, we use the information in the NodeShape to determine we should set inputCol/outputCol or inputCols/outputCols.

In MLeap side, when load mode, we will transformer
"input" -> "input0"
"output" -> "output0"  so that transformer can build correct input/output schema.

This approach implemented in two place.
1. [MultiInOutSparkOp](https://github.com/combust/mleap/pull/875/files#diff-1063c1657116e38e169540d372df2d8f191d7977c59ce0a175ebbf6bd1787e5d) responsible to provide correct port->parameter mapping for load and store
2. [MultiInOutMleapOp](https://github.com/combust/mleap/pull/875/files#diff-70364d73fb8a3a1b044507f1c2cba9a69c2c8fe5958a2a7d04b8c6d720bf21ed) responsible to convert single input/output to multiple input/output.

### CODE REVIEW

File: mleap-spark/src/main/scala/org/apache/spark/ml/bundle/ops/feature/BinarizerOp.scala [hunk 1]
```diff
@@ -24,35 +22,28 @@ class BinarizerOp extends SimpleSparkOp[Binarizer] with MultiInOutFormatSparkOp[
...
-        } else {
-          model.withValue("input_shapes", Value.dataShape(sparkToMleapDataShape(dataset.schema(obj.getInputCol), dataset)))
-        }
+      if(obj.isSet(obj.inputCols)) {
+        val inputShapes = obj.getInputCols.map(i => sparkToMleapDataShape(dataset.schema(i), dataset): DataShape)
+        model.withValue("input_shapes_list", Value.dataShapeList(inputShapes))
+          .withValue("thresholds", Value.doubleList(obj.getThresholds))
+      } else {
+        model.withValue("input_shapes", Value.dataShape(sparkToMleapDataShape(dataset.schema(obj.getInputCol), dataset)))
+          .withValue("threshold", Value.double(obj.getThreshold))
```
[the remark this answers is not in the document]
Comment (austinzh) [reply mleap-spark/src/main/scala/org/apache/spark/ml/bundle/ops/feature/BinarizerOp.scala:32]: Yes. Spark will report paramter not set

File: mleap-spark-base/src/main/scala/org/apache/spark/ml/bundle/MultiInOutFormatSparkOp.scala [hunk 2]
```diff
@@ -1,61 +1,53 @@
...
+import org.apache.spark.ml.param.shared._
+import ml.combust.bundle.op.OpModel
 
-trait MultiInOutFormatSparkOp[
-  N <: Transformer with HasInputCol with HasInputCols with HasOutputCol with HasOutputCols
-]{
+import scala.reflect.ClassTag
 
-  protected def saveMultiInOutFormat(model: Model, obj: N): Model = {
+abstract class MultiInOutOpModel[N <: Transformer with HasInputCol with HasInputCols with HasOutputCol with HasOutputCols] extends  OpModel[SparkBundleContext, N] {
```
[the remark this answers is not in the document]
Comment (austinzh) [reply mleap-spark-base/src/main/scala/org/apache/spark/ml/bundle/MultiInOutFormatSparkOp.scala]: I don't this project has it's scala formatter, I will use IDE formatter to format it.

File: mleap-spark/src/main/scala/org/apache/spark/ml/bundle/ops/feature/StopWordsRemoverOp.scala [hunk 3]
```diff
@@ -10,24 +10,23 @@ import org.apache.spark.ml.feature.StopWordsRemover
...
+        .withValue("case_sensitive", Value.boolean(obj.getCaseSensitive))
     }
 
     override def load(model: Model)
                      (implicit context: BundleContext[SparkBundleContext]): StopWordsRemover = {
-      val obj = new StopWordsRemover(uid = "").setStopWords(model.value("stop_words").getStringList.toArray).
-        setCaseSensitive(model.value("case_sensitive").getBoolean)
-      loadMultiInOutFormat(model, obj)
+     new StopWordsRemover(uid = "")
+       .setStopWords(model.value("stop_words").getStringList.toArray)
```
[the remark this answers is not in the document]
Comment (austinzh) [reply mleap-spark/src/main/scala/org/apache/spark/ml/bundle/ops/feature/StopWordsRemoverOp.scala:28]: No it won't, sparkLoad only do two thing.
1. clone and set uid
2. copy parameters from model to the clone
So here in the  val Model, it's the place to set all parameters that not columns and construct the transfomer(model)

File: mleap-spark-base/src/main/scala/org/apache/spark/ml/bundle/MultiInOutFormatSparkOp.scala [hunk 4]
```diff
@@ -1,61 +1,53 @@
...
   }
+  override def store(model: Model, obj: N)(implicit context: BundleContext[SparkBundleContext]): Model = {
+    validateParams(obj)
+    model
+  }
+}
+
+abstract class MultiInOutFormatSparkOp[
+  N <: Transformer with HasInputCol with HasInputCols with HasOutputCol with HasOutputCols
+](implicit ct: ClassTag[N]) extends  SimpleSparkOp[N] {
```
[the remark this answers is not in the document]
Comment (austinzh) [reply mleap-spark-base/src/main/scala/org/apache/spark/ml/bundle/MultiInOutFormatSparkOp.scala]: reformated

File: mleap-spark/src/main/scala/org/apache/spark/ml/bundle/ops/feature/StringIndexerOp.scala [hunk 5]
```diff
@@ -36,15 +37,14 @@ class StringIndexerOp extends SimpleSparkOp[StringIndexerModel] with MultiInOutF
...
-        for ( i <- 0 to label_length - 1) {
+        for ( i <- 0 until label_length) {
           collectedLabels(i) = model.value(s"labels_array_$i").getStringList.toArray
         }
         collectedLabels
       }
 
-      val obj = new StringIndexerModel(labelsArray = labelsArray).
+      new StringIndexerModel(labelsArray = labelsArray).
         setHandleInvalid(model.value("handle_invalid").getString)
```
[the remark this answers is not in the document]
Comment (austinzh) [reply mleap-spark/src/main/scala/org/apache/spark/ml/bundle/ops/feature/StringIndexerOp.scala:47]: reformated

**Review by ltrottier-yelp — APPROVED** [review]
LGTM

**Pull request merged**